RGBD failing to compile with OpenGL enabled

##### System information (version)
<!-- Example
- OpenCV => 3.1
- Operating System / Platform => Windows 64 Bit
- Compiler => Visual Studio 2015
-->

- OpenCV => Latest master
- Operating System / Platform => Windows 10 x64
- Compiler => Visual Studio 2017 x64

##### Detailed description

With OpenGL enabled and RGBD, it does not compile due to an error:
```
OpenCV-Contrib\modules\rgbd\src\dynafu.cpp(234): error C2131: expression did not evaluate to a constant
OpenCV-Contrib\modules\rgbd\src\dynafu.cpp(234): note: failure was caused by a read of a variable outside its lifetime
OpenCV-Contrib\modules\rgbd\src\dynafu.cpp(234): note: see usage of 'this'
OpenCV-Contrib\modules\rgbd\src\dynafu.cpp(206): note: while compiling class template member function 'void cv::dynafu::DynaFuImpl<cv::Mat>::drawScene(cv::OutputArray,cv::OutputArray)'
OpenCV-Contrib\modules\rgbd\src\dynafu.cpp(526): note: see reference to function template instantiation 'void cv::dynafu::DynaFuImpl<cv::Mat>::drawScene(cv::OutputArray,cv::OutputArray)' being compiled
OpenCV-Contrib\modules\rgbd\src\dynafu.cpp(275): note: while compiling class template member function 'cv::dynafu::DynaFuImpl<cv::Mat>::~DynaFuImpl(void)'
OpenCV-Contrib\modules\rgbd\src\dynafu.cpp(291): note: see reference to class template instantiation 'cv::dynafu::DynaFuImpl<cv::Mat>' being compiled
OpenCV-Contrib\modules\rgbd\src\dynafu.cpp(235): error C2131: expression did not evaluate to a constant
OpenCV-Contrib\modules\rgbd\src\dynafu.cpp(235): note: failure was caused by a read of a variable outside its lifetime
OpenCV-Contrib\modules\rgbd\src\dynafu.cpp(235): note: see usage of 'this'
```

The exact code that is failing is here:
```
float f[params.frameSize.width*params.frameSize.height];
```

##### Steps to reproduce

Compile with OpenGL and RGBD enabled.

Problem is not easy to solve. You can use vector<float> instaed of float[] but it does not solve problem. New problem is glext.h which is not available on windows. You can download it with khrplatform.h and then you will have link errors :

1>dynafu.obj : error LNK2019: unresolved external symbol glBindRenderbufferEXT referenced in function "public: __cdecl cv::dynafu::DynaFuImpl<class cv::Mat>::DynaFuImpl<class cv::Mat>(struct cv::dynafu::Params const &)" (??0?$DynaFuImpl@VMat@cv@@@dynafu@cv@@QEAA@AEBUParams@12@@Z)
1>dynafu.obj : error LNK2019: unresolved external symbol glGenRenderbuffersEXT referenced in function "public: __cdecl cv::dynafu::DynaFuImpl<class cv::Mat>::DynaFuImpl<class cv::Mat>(struct cv::dynafu::Params const &)" (??0?$DynaFuImpl@VMat@cv@@@dynafu@cv@@QEAA@AEBUParams@12@@Z)
1>dynafu.obj : error LNK2019: unresolved external symbol glRenderbufferStorageEXT referenced in function "public: __cdecl cv::dynafu::DynaFuImpl<class cv::Mat>::DynaFuImpl<class cv::Mat>(struct cv::dynafu::Params const &)" (??0?$DynaFuImpl@VMat@cv@@@dynafu@cv@@QEAA@AEBUParams@12@@Z)
1>dynafu.obj : error LNK2019: unresolved external symbol glBindFramebufferEXT referenced in function "public: __cdecl cv::dynafu::DynaFuImpl<class cv::Mat>::DynaFuImpl<class cv::Mat>(struct cv::dynafu::Params const &)" (??0?$DynaFuImpl@VMat@cv@@@dynafu@cv@@QEAA@AEBUParams@12@@Z)
1>dynafu.obj : error LNK2019: unresolved external symbol glGenFramebuffersEXT referenced in function "public: __cdecl cv::dynafu::DynaFuImpl<class cv::Mat>::DynaFuImpl<class cv::Mat>(struct cv::dynafu::Params const &)" (??0?$DynaFuImpl@VMat@cv@@@dynafu@cv@@QEAA@AEBUParams@12@@Z)
1>dynafu.obj : error LNK2019: unresolved external symbol glFramebufferRenderbufferEXT referenced in function "public: __cdecl cv::dynafu::DynaFuImpl<class cv::Mat>::DynaFuImpl<class cv::Mat>(struct cv::dynafu::Params const &)" (??0?$DynaFuImpl@VMat@cv@@@dynafu@cv@@QEAA@AEBUParams@12@@Z)


This module is not compatible with windows and msvc. It is not available in 3.4 [branch](https://github.com/opencv/opencv_contrib/tree/3.4/modules/rgbd/src)

I have encountered what seems to be a related problem, when compiling under Linux.

When OPENCV_ENABLE_NONFREE is set, and OpenGL_GL_PREFERENCE is set to GLVND (newer OpenGL), the build fails, with reference to libopencv_rgbd:

../../lib/libopencv_rgbd.so.4.2.0: undefined reference to `glRenderbufferStorageEXT'
etc...

But if OpenGL_GL_PREFERENCE is set to LEGACY, there is no error.

--- a/modules/rgbd/src/dynafu.cpp
+++ b/modules/rgbd/src/dynafu.cpp
@@ -25,6 +25,54 @@
 #endif
 # include <GL/gl.h>
 #endif
+
+// GL Extention definitions missing from standard Win32 gl.h
+#if defined(_WIN32) && !defined(GL_RENDERBUFFER_EXT)
+#define GL_COLOR_ATTACHMENT0_EXT 0x8CE0
+#define GL_DEPTH_ATTACHMENT_EXT 0x8D00
+#define GL_FRAMEBUFFER_EXT 0x8D40
+#define GL_RENDERBUFFER_EXT 0x8D41
+namespace {
+PROC _wglGetProcAddress(const char *name)
+{
+  auto proc = wglGetProcAddress(name);
+  if (!proc)
+    CV_Error(cv::Error::OpenGlApiCallError, cv::format("Can't load OpenGL extension [%s]", name) );
+  return proc;
+}
+
+void glGenFramebuffersEXT(GLsizei n, GLuint *framebuffers)
+{
+  static auto proc = reinterpret_cast<void(*)(GLsizei, GLuint*)>(_wglGetProcAddress(__func__));
+  proc(n, framebuffers);
+}
+void glGenRenderbuffersEXT(GLsizei n, GLuint *renderbuffers)
+{
+  static auto proc = reinterpret_cast<void(*)(GLsizei, GLuint*)>(_wglGetProcAddress(__func__));
+  proc(n, renderbuffers);
+}
+void glBindRenderbufferEXT(GLenum target, GLuint renderbuffer)
+{
+  static auto proc = reinterpret_cast<void(*)(GLenum, GLuint)>(_wglGetProcAddress(__func__));
+  proc(target, renderbuffer);
+}
+void glBindFramebufferEXT(GLenum target, GLuint framebuffer)
+{
+  static auto proc = reinterpret_cast<void(*)(GLenum, GLuint)>(_wglGetProcAddress(__func__));
+  proc(target, framebuffer);
+}
+void glFramebufferRenderbufferEXT(GLenum target, GLenum attachment, GLenum renderbuffertarget, GLuint renderbuffer)
+{
+  static auto proc = reinterpret_cast<void(*)(GLenum, GLenum, GLenum, GLuint)>(_wglGetProcAddress(__func__));
+  proc(target, attachment, renderbuffertarget, renderbuffer);
+}
+void glRenderbufferStorageEXT(GLenum target, GLenum internalformat, GLsizei width, GLsizei height)
+{
+  static auto proc = reinterpret_cast<void(*)(GLenum, GLenum, GLsizei, GLsizei)>(_wglGetProcAddress(__func__));
+  proc(target, internalformat, width, height);
+}
+} // anonymous namespace
+#endif // defined(_WIN32) && !defined(GL_RENDERBUFFER_EXT)
 #else
 # define NO_OGL_ERR CV_Error(cv::Error::OpenGlNotSupported, \
                     "OpenGL support not enabled. Please rebuild the library with OpenGL support");

